### PR TITLE
Extend using statement

### LINKED ISSUE #9211: Extend "using for"
This is a proposal to extend the "using for" directive in the following ways:

 - allow it at file level to be valid for all contracts and functions in that file
 - allow its first argument to be a free function
 - allow its first argument to be an imported module
 - allow its first argument to be `*`

The semantics are as follows:

The directive `using * for ...` matches all free functions in the current file (including non-scoped free functions imported from somewhere else), but it does not match functions in libraries or functions imported scoped into a module (`moduleName.functionName`).

The directive `using functionName for ...` matches the specified function. Here, `functionName` can be a path like `moduleName.functionName`.

The directive `using moduleName for ...` matches all free functions in the imported module (via `import "fileName" as moduleName`) including free functions imported from that module at file level. Here, `moduleName` can be a path like `moduleA.moduleB`.

Note that most of the new uses are incompatible with flatteners.

---
Update (2021-09-20):

Extend the `using A for B` statement everywhere to allow `A` to also be
 - a list of file-level functions in curly braces.
 - the name of a module
 - a single file-level function
 - `*`

If used at file-level, `B` cannot be `*`.

If used at file-level also allow `export using A for B`, when `B` is a type declared at file-level in the current file.

Clarification: In the above, all "items" can be paths (`A.B.f`) and don't have to be single identifiers.

### PR DESCRIPTION
**Pull request by hrkrshnn**
Fixes https://github.com/ethereum/solidity/issues/9211

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -3430,6 +3431,57 @@ void TypeChecker::endVisit(Literal const& _literal)
 
 void TypeChecker::endVisit(UsingForDirective const& _usingFor)
 {
+	auto checkFunctionDefinition = [&](FunctionDefinition const& _functionDefinition, SourceLocation _location)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
	auto checkFunctionCompatibility = [&](FunctionDefinition const& _functionDefinition, SourceLocation _location)
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -3430,6 +3431,57 @@ void TypeChecker::endVisit(Literal const& _literal)
...
+				"The function \"" + _functionDefinition.name() + "\" "+
+				"cannot be bound to the type \"" + _usingFor.typeName()->annotation().type->toString() +
+				"\" because the type cannot be implicitly converted to the first argument" +
+				" of the function (\"" + functionType->selfType()->toString() + "\")."
+			);
+	};
+
+	UsingForDirective::LHS const& lhs = _usingFor.lhs();
+	std::visit(util::GenericVisitor{
+		[&](UsingForDirective::LibraryOrFunctionOrModule const& _libraryOrFunctionOrModule) {
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Maybe we could have a helper function in `UsingForDirective` that allows us to handle `using f for X` in the same way as `using {f} for X` if `f` is a function?

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -3430,6 +3431,57 @@ void TypeChecker::endVisit(Literal const& _literal)
...
+			Declaration const* declaration = _libraryOrFunctionOrModule.name->annotation().referencedDeclaration;
+			if (auto functionDefinition = dynamic_cast<FunctionDefinition const*>(declaration))
+				checkFunctionDefinition(*functionDefinition, _libraryOrFunctionOrModule.name->location());
+		},
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: If it's a library or module we don't check (which is fine I think, just wanted to confirm)?

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 4]
```diff
@@ -453,12 +454,54 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
...
 
-	if (!library || !library->isLibrary())
-		m_errorReporter.fatalTypeError(4357_error, _usingFor.libraryName().location(), "Library name expected.");
+	auto checkFunctionDefinition = [&](FunctionDefinition const* _functionDefinition, SourceLocation _location)
+	{
+		if (!_functionDefinition->isFree())
+			m_errorReporter.typeError(
+				4167_error,
+				_location,
+				"Only free functions can be bound to a type in a \"using\" statement"
```
Comment (chriseth) [inline libsolidity/analysis/DeclarationTypeChecker.cpp]:
```suggestion
				"Functions need to be free in order to bind them explicitly to a type in a \"using\" statement"
```

Comment (chriseth) [reply libsolidity/analysis/DeclarationTypeChecker.cpp]: But mabe it's not really better :)

File: libsolidity/ast/Types.cpp [hunk 5]
```diff
@@ -363,21 +388,68 @@ MemberList::MemberMap Type::boundFunctions(Type const& _type, ASTNode const& _sc
...
+					FunctionTypePointer fun = dynamic_cast<FunctionType const&>(*function.type()).asBoundFunction();
+					solAssert(fun && _type.isImplicitlyConvertibleTo(*fun->selfType()) , "");
+
+					addMember(&function, fun);
+				}
+			},
+			[&](UsingForDirective::Asterisk const&)
+			{
+				// TODO change this when using statement is allowed at file level.
+				ContractDefinition const& contract = dynamic_cast<ContractDefinition const&>(_scope);
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Shouldn't `*` only capture elements from the current scope? I would actually disallow except when used at file level.

**Comment by chriseth** [comment]
Looks good in general. I think there is some code duplication and we should also check if everything works as expected when the name of a function is changed in an `import` statement.

File: libsolidity/ast/Types.cpp (hunk 5 above)
Comment (hrkrshnn) [reply libsolidity/ast/Types.cpp]:
```solidity
library L {
    function f(S memory s) internal {}
}
struct S { uint x; }
contract C {
    using L for *;
    function g(S memory s) external {
        s.f();
    }
}
```
The following works. So I think capturing from the current source unit should be okay.

File: test/libsolidity/semanticTests/using/module_2.sol [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        return x.id();
+    }
+    function g(uint x) public pure returns (uint) {
+        return x.one_ext();
+    }
+}
+// ====
+// compileViaYul: also
+// ----
+// library: L
```
Comment (hrkrshnn) [inline test/libsolidity/semanticTests/using/module_2.sol]: @cameel This fails currently, I think it is expecting `A.L`. However, isoltest seems to not support separating with dots.

File: test/libsolidity/syntaxTests/using/module_4.sol [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    return x;
+}
+
+==== Source: B ====
+import "A";
+
+contract C {
+    // This only attaches the free functions in the current source unit.
+    // Therefore, `id` is not accessible.
+    using * for uint;
```
Comment (hrkrshnn) [inline test/libsolidity/syntaxTests/using/module_4.sol]: @chriseth Is this how we want the semantics?

File: test/libsolidity/semanticTests/using/module_2.sol (hunk 6 above)
Comment (cameel) [reply test/libsolidity/semanticTests/using/module_2.sol]: ~True, but I think it was just never implemented (rather than broken by #12073). The thing that the PR changed is that you can now put the source unit name in front of the library name.~

~Should not be hard to implement though. Changes are likely needed in the same places as in #12073. The main thing was the [change in `TestFileParser.cpp`](https://github.com/soroosh-sdi/solidity/commit/816d8021e4f57b1b32eb3ce1fbfa421a585a0f0f#diff-ca8456220457a41b507d3968dbffa4ed8b777a91653fec0eedd4985fd9bf5606). The rest is mostly adjusting tests.~

File: test/libsolidity/syntaxTests/using/module_4.sol (hunk 7 above)
Comment (chriseth) [reply test/libsolidity/syntaxTests/using/module_4.sol]: I would suggest that these things are only valid at file level.

Comment (chriseth) [reply test/libsolidity/syntaxTests/using/module_4.sol]: Especially inside contracts it is horribly difficult due to inheritance.

File: test/libsolidity/semanticTests/using/module_2.sol (hunk 6 above)
Comment (cameel) [reply test/libsolidity/semanticTests/using/module_2.sol]: Oh, wait. Even if you do `import "A" as M` and refer to library `L` as `M.L`, its fully qualified name does not become `M.L`. It's always qualified with the source unit name. So on the CLI you would use `A:L`. In soltest you need ~`"A".L`~ `"A":L` because source unit names can contain arbitrary characters and it was too hard to get the parser to allow that without something like quoting or a special prefix.

So no need to handle dots at all.

Comment (hrkrshnn) [reply test/libsolidity/semanticTests/using/module_2.sol]: > So no need to handle dots at all.

What do you mean? Changing to "A".L is also unsupported by isoltest currently.

Comment (cameel) [reply test/libsolidity/semanticTests/using/module_2.sol]: Sorry, it's a typo. I meant `"A":L`.

File: test/libsolidity/syntaxTests/using/module_4.sol (hunk 7 above)
Comment (hrkrshnn) [reply test/libsolidity/syntaxTests/using/module_4.sol]: Disallowed this for contracts.

**Comment by hrkrshnn** [comment]
There is an annoying problem here. One difference between libraries and free functions in using for statement is that, unlike `using LibraryName for T`, `using free_function for T` can be immediately type checked.

But this has issues for
```solidity
function sum(uint[] calldata arr) returns (uint partialSum) {
    for(uint i = 0; i < arr.length; i++) {
        partialSum += arr[i];
    }
}
contract C {
    using sum for uint[];
    function f(uint[] memory arr) external returns (uint) {
        arr.sum();
    }
}
```
The underlying type `uint[]` has to be treated as a storage type (the default location?). But at instantiation time, we need a conversion from `memory -> calldata`, which is not possible. There are two ways to proceed here:

If we really want `using X for T;` to be type checked at declaration time, then we should allow `using f for uint[] calldata;`
The other option is to individually type check during the calling, which doesn't look nice.

What do you think @chriseth

**Comment by chriseth** [comment]
What about doing type checking right away for non-reference types and delaying it if there is at least one storage location that works?

**Comment by hrkrshnn** [comment]
> What about doing type checking right away for non-reference types and delaying it if there is at least one storage location that works?

I could do that.

> it if there is at least one storage location that works?

This might be tricky, but I can try.

File: libsolidity/ast/AST.h [hunk 8]
```diff
@@ -629,33 +631,71 @@ class InheritanceSpecifier: public ASTNode
...
+	};
+	/// `using {f1, f2, ..., fn} for T;`
+	struct FunctionList
+	{
+		std::vector<ASTPointer<IdentifierPath>> functions;
+	};
+	/// `using * for T;`
+	struct Asterisk {};
+
+	using LHS = std::variant<LibraryOrFunctionOrModule, FunctionList, Asterisk>;
```
Comment (chriseth) [inline libsolidity/ast/AST.h]: I think we should try to find a better name for this.

File: test/libsolidity/semanticTests/using/module_1.sol [hunk 9]
```diff
@@ -0,0 +1,18 @@
+==== Source: A ====
+function id(uint x) pure returns (uint) {
+    return x;
+}
+
+==== Source: B ====
+import "A" as M;
+contract C {
+    using M for uint;
+	function f(uint x) public returns (uint) {
```
Comment (bshastry) [inline test/libsolidity/semanticTests/using/module_1.sol]:
```suggestion
    function f(uint x) public returns (uint) {
```

File: docs/grammar/SolidityParser.g4 [hunk 10]
```diff
@@ -314,7 +314,7 @@ errorDefinition:
  * Using directive to bind library functions to types.
  * Can occur within contracts and libraries.
  */
-usingDirective: Using identifierPath For (Mul | typeName) Semicolon;
+usingDirective: Using (identifierPath | (LBrace identifierPath (Comma identifierPath)* RBrace) | Mul) For (Mul | typeName) Semicolon;
```
Comment (bshastry) [inline docs/grammar/SolidityParser.g4]: `usingDirective` is missing from sourceUnit definitions i.e., this PR also enables `using * for T` at source unit level, right?

File: test/libsolidity/syntaxTests/using/module_2.sol [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+function id(uint x) pure returns (uint) {
+    return x;
+}
+
+==== Source: B ====
+import {id as Id} from "A";
+
+contract C {
+    using Id for uint;
+	function f(uint x) public pure returns (uint) {
```
Comment (bshastry) [inline test/libsolidity/syntaxTests/using/module_2.sol:11]:
```suggestion
    function f(uint x) public pure returns (uint) {
```

File: test/libsolidity/syntaxTests/using/module_3.sol [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+function id(uint x) pure returns (uint) {
+    return x;
+}
+
+==== Source: B ====
+import "A" as M;
+
+contract C {
+    using M.id for uint;
+	function f(uint x) public pure returns (uint) {
```
Comment (bshastry) [inline test/libsolidity/syntaxTests/using/module_3.sol]:
```suggestion
    function f(uint x) public pure returns (uint) {
```

File: test/libsolidity/semanticTests/using/module_1.sol [hunk 13]
```diff
@@ -0,0 +1,18 @@
+==== Source: A ====
+function id(uint x) pure returns (uint) {
+    return x;
+}
+
+==== Source: B ====
+import "A" as M;
+contract C {
+    using M for uint;
```
Comment (bshastry) [inline test/libsolidity/semanticTests/using/module_1.sol]: If the import statement above looks like `import "A.sol";` (`A.sol` being the file instead of simply `A`) and then we do `using A.sol for uint`, does the parser parse it as intended?

In other words, if no module name is defined, does it default the module name default to file name?

File: docs/grammar/SolidityParser.g4 (hunk 10 above)
Comment (chriseth) [reply docs/grammar/SolidityParser.g4]: Right, maybe the is an argument for not having this logic in the parser :)

**Comment by chriseth** [comment]
The current implementation performs type checking at the `using for` place when the overload is unique and not a module. Type checking is delayed to the call location if there are overloads.

**Comment by chriseth** [comment]
The above is wrong: `using {f, g} for uint;` requires `f` and `g` to be unique.

File: test/libsolidity/semanticTests/using/module_1.sol (hunk 13 above)
Comment (chriseth) [reply test/libsolidity/semanticTests/using/module_1.sol]: No, it does not work like that at all. `import "A.sol";` imports all top-level elements from `"A.sol"` into the current scope, while `import "A.sol" as M;` creates a new top-level scope called `M` that you can reference.

Comment (chriseth) [reply test/libsolidity/semanticTests/using/module_1.sol]: In the second syntax you need to access items as `M.f` while in the first you just use `f`.

**Comment by chriseth** [comment]
Currently, `using {f} for x` fails if `f` is overloaded. Do we want to change that?

Also @cameel thinks that `using X ...` and `using {X} ...` should be fully equivalent, and I kind of agree.

**Comment by chriseth** [comment]
Instead of `export using ...` we could use `using {f, h} for MyType sticky;`

**Comment by chriseth** [comment]
Having thought about this, I believe a simpler specification / implementation might make sense for now:
- If `{...}` is used, it can only contain a list of functions (but they can be library functions)
- I would drop `using ModuleName for ...`
- `using LibraryName` is not allowed at file level (we can change that later)
- If `using { ... } for MyType` is used at file level and `MyType` (struct, enum or type) is defined in the same file at top level, the functions are available everywhere the type is available - if we can come up with a good keyword, this can still be changed

In summary, the change to how it was before this PR:
 - add `using {...} for ExplicitType` (no `*`)
 - allow `using` at file-level, but only in the `{...}`-form and make it auto-attach

We can even think about disallowing `using` at file-level on types that are not defined there.

**Comment by chriseth** [comment]
We have to check what happens when a virtual function is referenced.

**Comment by chriseth** [comment]
virtual functions are of course not a problem because they cannot be referenced: The functions have to be free or library functions and this is already checked.

**Comment by chriseth** [comment]
Please review - the first commit is a more complicated version.

File: libsolidity/ast/AST.h [hunk 14]
```diff
@@ -629,33 +630,47 @@ class InheritanceSpecifier: public ASTNode
...
+ *     `fn`, respectively to `T`.
+ *
+ * For version 3, T has to be implicitly convertible to the first parameter type of
+ * all functions, and this is checked at the point of the using statement. For versions 1 and
+ * 2, this check is only done when a function is called.
  */
 class UsingForDirective: public ASTNode
 {
 public:
+	struct Asterisk {};
```
Comment (ekpyron) [inline libsolidity/ast/AST.h]: Is this used?

**Comment by ekpyron** [comment]
Do we have a test case for using a function multiple times in the list? ``using {f, f} for``?

File: test/libsolidity/syntaxTests/using/free_reference_type.sol [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+}
+function g(uint[] storage x) pure returns (uint) {
+    return x[0];
+}
+function h(uint[] calldata x) pure returns (uint) {
+    return x[0];
+}
+using {f, g, h} for uint[];
+// ----
+// TypeError 2527: (131-135): Function declared as pure, but this expression (potentially) reads from the environment or state and thus requires "view".
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/using/free_reference_type.sol]: Is it the intention of this test to have this error?

File: test/libsolidity/semanticTests/using/imported_functions.sol [hunk 16]
```diff
@@ -0,0 +1,19 @@
+==== Source: A ====
+function inc(uint x) pure returns (uint) {
+    return x + 1;
+}
+
+==== Source: B ====
+import {inc as f} from "A";
+import "A";
+using * for uint;
```
Comment (ekpyron) [inline test/libsolidity/semanticTests/using/imported_functions.sol]: This is neither in the grammar nor in the documentation, is it?
Even more so, is this really something we expect to be used :-)?

File: docs/grammar/SolidityParser.g4 [hunk 17]
```diff
@@ -311,10 +311,10 @@ errorDefinition:
 	Semicolon;
 
 /**
- * Using directive to bind library functions to types.
- * Can occur within contracts and libraries.
+ * Using directive to bind library functions and free functions to types.
+ * Can occur within contracts and libraries and at the file level.
```
Comment (ekpyron) [inline docs/grammar/SolidityParser.g4:316]: Doesn't it need to be added to ``sourceUnit`` then?

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 18]
```diff
@@ -451,12 +452,39 @@ void DeclarationTypeChecker::endVisit(VariableDeclaration const& _variable)
...
+					dynamic_cast<ContractDefinition const*>(functionDefinition->scope())->isLibrary()
+				))
+					m_errorReporter.typeError(
+						4167_error,
+						function->location(),
+						"Only free functions and library functions can be bound to a type in a \"using\" statement"
+					);
+			}
+			else
+				m_errorReporter.fatalTypeError(8187_error, function->location(), "Expected function name." );
```
Comment (ekpyron) [inline libsolidity/analysis/DeclarationTypeChecker.cpp:471]: CI complains that this is untested.

File: test/libsolidity/syntaxTests/using/using_contract_err.sol [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+}
+interface I {
+    function f(uint) external;
+}
+
+contract Test {
+    using C for uint;
+}
+// ----
+// TypeError 4357: (127-128): Library name expected. If you want to attach a function, use '{...}'.
```
Comment (ekpyron) [inline test/libsolidity/syntaxTests/using/using_contract_err.sol:13]: ~~If we want to be perfectionistic, we can not emit the second part inside a contract... but also fine.~~ Nevermind

File: libsolidity/analysis/TypeChecker.cpp [hunk 20]
```diff
@@ -3586,12 +3587,59 @@ void TypeChecker::endVisit(Literal const& _literal)
...
+	}
+
+	for (ASTPointer<IdentifierPath> const& path: _usingFor.functionsOrLibrary())
+	{
+		solAssert(path->annotation().referencedDeclaration);
+		// No type checking if it is a library.
+		FunctionDefinition const* functionDefinition =
+			dynamic_cast<FunctionDefinition const*>(path->annotation().referencedDeclaration);
+		if (!functionDefinition)
+			continue;
```
Comment (ekpyron) [inline libsolidity/analysis/TypeChecker.cpp]: I'd consider to instead do something like
```
if (!_usingFor.usesBraces())
{
  // assert that _usingFor.functionsOrLibrary() is a single library
  return;
}
```
Before the loop and then assert that ``functionDefinition`` is not ``nullptr`` here.

But also fine as is.

File: libsolidity/ast/Types.cpp [hunk 21]
```diff
@@ -362,20 +381,28 @@ MemberList::MemberMap Type::boundFunctions(Type const& _type, ASTNode const& _sc
 			)
 		)
 			continue;
-		auto const& library = dynamic_cast<ContractDefinition const&>(
-			*ufd->libraryName().annotation().referencedDeclaration
-		);
-		for (FunctionDefinition const* function: library.definedFunctions())
+
```
Comment (ekpyron) [inline libsolidity/ast/Types.cpp:385]: Here I also wonder if checking ``ufd->usesBraces()`` and asserting accordingly would be more fool-proof.
But yeah - on the other hand "usesBraces" is a syntactic property and it makes sense not to consider it anymore later on - resp. again: also fine as is.

**Review by ekpyron — COMMENTED** [review]
Looks good! But there seems to be some issue with the docs CI run. It could also use a rebase to fix unrelated CI issues I think.

File: libsolidity/analysis/TypeChecker.cpp (hunk 20 above)
Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: Restructured it a bit.

File: libsolidity/analysis/TypeChecker.cpp [hunk 22]
```diff
@@ -3586,12 +3587,68 @@ void TypeChecker::endVisit(Literal const& _literal)
...
+	Type const* normalizedType = TypeProvider::withLocationIfReference(
+		DataLocation::Storage,
+		_usingFor.typeName()->annotation().type
+	);
+	solAssert(normalizedType);
+
+	for (ASTPointer<IdentifierPath> const& path: _usingFor.functionsOrLibrary())
+	{
+		solAssert(path->annotation().referencedDeclaration);
+		// No type checking if it is a library.
```
Comment (ekpyron) [inline libsolidity/analysis/TypeChecker.cpp]: The comment should be removed now - but apart from that the new version looks much nicer.

File: docs/contracts/using-for.rst [hunk 23]
```diff
@@ -6,71 +6,89 @@
...
+
+If you specify a library, *all* functions in the library are attached,
 even those where the type of the first parameter does not
 match the type of the object. The type is checked at the
 point the function is called and function overload
 resolution is performed.
 
+If you use a list of functions (``using {f, g, h, L.t} for uint;``),
+then the type (``uint``) has to be implicitly convertible to the
+first parameter of each of these functions. This mean the check is
```
Comment (ekpyron) [inline docs/contracts/using-for.rst]:
```suggestion
first parameter of each of these functions. This check is
```
or otherwise ``means``

File: docs/contracts/using-for.rst [hunk 24]
```diff
@@ -6,71 +6,89 @@
...
-contract, including within all of its functions, and has no effect
-outside of the contract in which it is used. The directive
-may only be used inside a contract, not inside any of its functions.
+scope (either the contract or the current module/source unit),
+including within all of its functions, and has no effect
+outside of the contract or module in which it is used.
 
 Let us rewrite the set example from the
-:ref:`libraries` in this way:
+:ref:`libraries` in this way, using file-level functions
```
Comment (ekpyron) [inline docs/contracts/using-for.rst]:
```suggestion
:ref:`libraries` section in this way, using file-level functions
```
Or otherwise the ``the`` before is weird, isn't it?

File: docs/contracts/using-for.rst [hunk 25]
```diff
@@ -6,71 +6,89 @@
...
-    pragma solidity >=0.6.0 <0.9.0;
-
+    pragma solidity ^0.8.12;
 
-    // This is the same code as before, just without comments
     struct Data { mapping(uint => bool) flags; }
+    // Now we attach functions to the type.
+    // This is valid in the rest of the module.
+    // If you import the module, you have to
+    // repeat the statement there, potentially as
```
Comment (ekpyron) [inline docs/contracts/using-for.rst]:
```suggestion
    // repeat the using statement there, for example as
```

File: docs/contracts/using-for.rst [hunk 26]
```diff
@@ -6,71 +6,89 @@
...
 
     // SPDX-License-Identifier: GPL-3.0
-    pragma solidity >=0.6.0 <0.9.0;
-
+    pragma solidity ^0.8.12;
 
-    // This is the same code as before, just without comments
     struct Data { mapping(uint => bool) flags; }
+    // Now we attach functions to the type.
+    // This is valid in the rest of the module.
```
Comment (ekpyron) [inline docs/contracts/using-for.rst]:
```suggestion
    // The attached functions can be used throughout the rest of the module.
```

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 27]
```diff
@@ -310,10 +311,16 @@ bool ASTJsonConverter::visit(InheritanceSpecifier const& _node)
 
 bool ASTJsonConverter::visit(UsingForDirective const& _node)
 {
-	setJsonNode(_node, "UsingForDirective", {
-		make_pair("libraryName", toJson(_node.libraryName())),
+	std::vector<pair<string, Json::Value>> attributes = {
```
Comment (ekpyron) [inline libsolidity/ast/ASTJsonConverter.cpp]:
```suggestion
	vector<pair<string, Json::Value>> attributes = {
```

File: libsolidity/ast/Types.cpp [hunk 28]
```diff
@@ -330,30 +331,48 @@ Type const* Type::fullEncodingType(bool _inLibraryCall, bool _encoderV2, bool) c
...
 	// Normalise data location of type.
 	DataLocation typeLocation = DataLocation::Storage;
 	if (auto refType = dynamic_cast<ReferenceType const*>(&_type))
 		typeLocation = refType->location();
 
-	set<Declaration const*> seenFunctions;
+	set<pair<string, Declaration const*>> seenFunctions;
 	MemberList::MemberMap members;
 
+	auto addFunction = [&](FunctionDefinition const& _function, optional<string> _name = {})
```
Comment (ekpyron) [inline libsolidity/ast/Types.cpp:353]:
```suggestion
	auto addFunction = [&, seenFunctions = set<pair<string, Declaration const*>>{}](FunctionDefinition const& _function, optional<string> _name = {})
```
I'd actually consider this - I was wondering where it's actually used for a second, since it's defined outside.
But also fine keeping it as is.

**Review by ekpyron — COMMENTED** [review]
Only more optional docs and style comments. Apart from maybe the outdated comment in https://github.com/ethereum/solidity/pull/12121/files#r801814258 I'd be fine with keeping as is.

**Comment by ekpyron** [comment]
Oh, and it probably wouldn't hurt to have a test for importing a file that uses a global ``using``, checking that the attached functions are no longer available after import.

File: libsolidity/ast/Types.cpp (hunk 28 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp:353]: I tried it, but it didn't work. I moved the variable closer to the function, though.

Comment (ekpyron) [reply libsolidity/ast/Types.cpp:353]: Ah, the lambda would probably have to be declared ``mutable`` to get a non-const capture. But yeah, doesn't matter.

File: libsolidity/parsing/Parser.cpp [hunk 29]
```diff
@@ -962,7 +965,26 @@ ASTPointer<UsingForDirective> Parser::parseUsingDirective()
 	ASTNodeFactory nodeFactory(*this);
 
 	expectToken(Token::Using);
-	ASTPointer<IdentifierPath> library(parseIdentifierPath());
+
+	vector<ASTPointer<IdentifierPath>> functions;
+	bool const usesBraces = m_scanner->currentToken() == Token::LBrace;
+	if (usesBraces)
+	{
+		vector<ASTPointer<IdentifierPath>> functionList;
```
Comment (ekpyron) [inline libsolidity/parsing/Parser.cpp]: Not sure this helps much compared to just adding to ``functions`` directly...

**Review by ekpyron — DISMISSED** [review]
Looks good to me now!

File: libsolidity/parsing/Parser.cpp (hunk 29 above)
Comment (chriseth) [reply libsolidity/parsing/Parser.cpp]: Oh, I think I even changed that, but it seems the change got lost.

File: libsolidity/parsing/Parser.cpp [hunk 30]
```diff
@@ -962,7 +965,26 @@ ASTPointer<UsingForDirective> Parser::parseUsingDirective()
...
+		advance();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: Wait - can't this just be moved inside the loop and the other one deleted?

**Review by ekpyron — DISMISSED** [review]

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 31]
```diff
@@ -310,10 +311,16 @@ bool ASTJsonConverter::visit(InheritanceSpecifier const& _node)
...
 bool ASTJsonConverter::visit(UsingForDirective const& _node)
 {
-	setJsonNode(_node, "UsingForDirective", {
-		make_pair("libraryName", toJson(_node.libraryName())),
+	vector<pair<string, Json::Value>> attributes = {
 		make_pair("typeName", _node.typeName() ? toJson(*_node.typeName()) : Json::nullValue)
-	});
+	};
+	if (_node.usesBraces())
+		attributes.emplace_back("functionList", toJson(_node.functionsOrLibrary()));
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonConverter.cpp]: To be compatible with future extensions, this should be an object.

**Comment by chriseth** [comment]
When we have `using {..., f, ...} for T`, we should enforce that the first parameter of `f` is equal to T up to data location, because it is less confusing and we also will get fewer problems with operators.

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 32]
```diff
@@ -311,10 +312,25 @@ bool ASTJsonConverter::visit(InheritanceSpecifier const& _node)
...
 		make_pair("typeName", _node.typeName() ? toJson(*_node.typeName()) : Json::nullValue)
-	});
+	};
+	if (_node.usesBraces())
+	{
+		Json::Value functionList;
+		for (auto const& function: _node.functionsOrLibrary())
+		{
+			Json::Value functionNode;
+			functionNode["function"] = toJson(*function);
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonConverter.cpp:324]: The `function` here is needed because of `using {add as +} for MyType;`.

File: docs/contracts/using-for.rst [hunk 33]
```diff
@@ -82,12 +100,13 @@ Let us rewrite the set example from the
         }
     }
 
-It is also possible to extend elementary types in that way:
+It is also possible to extend elementary types in that way.
```
Comment (ekpyron) [inline docs/contracts/using-for.rst]: Strictly speaking, the array below is not an "elementary type", though, is it?
Not sure we need to change anything, though.

**Review by ekpyron — COMMENTED** [review]
Maybe a test for
```
function f(uint256 x, uint256 y) pure {}
function f(uint256 x, uint256[] y) pure {}
using { f } for uint256;
```
?
We intentionally don't allow that, right?
And I guess mainly because it'd be annoying to implement? My guess would be that this not being possible might be rather annoying in the long run... this makes it impossible to add overloaded functions to a type without libraries, right?

**Comment by chriseth** [comment]
I'm growing less and less fond of function overloading. Rust made a good decision to ban it altogether.

**Comment by ekpyron** [comment]
> I'm growing less and less fond of function overloading. Rust made a good decision to ban it altogether.

Fair enough - I'd still expect this to be annoying as long as there generally is function overloading - but yeah, at least a test case would be nice in any case.

**Review by ekpyron — APPROVED** [review]
